pyblish/plugin.py

```diff
@@ -462,10 +503,38 @@ def process(plugin, context, instance=None, action=None):
 
     """
 
-    if issubclass(plugin, (ContextPlugin, InstancePlugin)):
-        result = __explicit_process(plugin, context, instance, action)
-    else:
-        result = __implicit_process(plugin, context, instance, action)
+    if not context.get('profiling'):
+        context.data['profiling'] = {}
+
+    profiler = ProfilingTimer()
+    with profiler:
+        if issubclass(plugin, (ContextPlugin, InstancePlugin)):
+            result = __explicit_process(plugin, context, instance, action)
+        else:
+            result = __implicit_process(plugin, context, instance, action)
+
+    profiling_data = {
+        "runtime": profiler.runtime,
+    }
+
+    # collected using psutil module, None if it is not installed
+    try:
+        profiling_data['initial_memory_sage'] = profiler.initial_mem
+        profiling_data['final_memory_usage'] = profiler.final_mem
+    except AttributeError:
+        # no memory information collected
+        pass
+
+    context.data['profiling'][plugin] = profiling_data
+    profiling_data["snapshot"] = {
+        "timestamp": int(time.time()),
+        "context": json.dumps(context, default=lambda o: '<not serializable>')
+    }
+
+    profiling_path = os.path.join(tempfile.gettempdir(),
+                                  "{}-pyblish_profile.json".format(context.id))
+
+    append_to_json(profiling_data, profiling_path)
```

Comment:
Ah no, I don't think this is a good idea.

Why not let the user decide what to do with the profiling information?

```py
context = util.publish()

with open("somefile.json", "w") as f:
  json.dump(f, context.data["profiling"])
```

GUIs could get a hold of it via the context as well, inbetween each process if need be, but likely also as a post-process of sorts.

Reply:
So you think that it shouldn't save to json file implicitly and it's up to user to handle data in `context.data.get('statistics')` (lets call it that) afterward? Just checking if I understand it correctly.

Reply:
Yes, don't write files without the user's knowledge. In the vast majority of cases, the user won't care for those files anyway. Profiling is primarily useful for the plug-in author.